src/contracts/collections/positions.rs

```diff
@@ -46,54 +45,200 @@ impl Positions {
-            let position = *positions.get(index as usize).unwrap();
-            positions_length -= 1;
-            positions.remove(index as usize);
-            self.positions
-                .insert(account_id, &(positions_length, positions));
-            Ok(position)
+        let positions_length = self.get_length(account_id);
+
+        if index >= positions_length {
+            return Err(InvariantError::PositionNotFound);
+        }
+
+        let position = self.positions.get((account_id, index));
+
+        if index < positions_length - 1 {
+            let last_position = self
+                .positions
+                .take((account_id, positions_length - 1))
+                .unwrap();
+            self.positions.insert((account_id, index), &last_position);
         } else {
-            Err(InvariantError::PositionNotFound)
+            self.positions.remove((account_id, index));
         }
+
+        self.positions_length
+            .insert(account_id, &(positions_length - 1));
+
+        Ok(position.unwrap())
     }
 
     pub fn transfer(
         &mut self,
         account_id: AccountId,
         index: u32,
-        receiver: AccountId,
+        receiver_account_id: AccountId,
     ) -> Result<(), InvariantError> {
         let position = self.remove(account_id, index)?;
-        self.add(receiver, position);
+        self.add(receiver_account_id, position);
 
         Ok(())
     }
 
     pub fn get_all(&self, account_id: AccountId) -> Vec<Position> {
-        self.positions.get(&account_id).unwrap_or_default().1
+        let mut positions = vec![];
+
+        for index in 0..self.get_length(account_id) {
+            let position = self.positions.get((account_id, index)).unwrap();
+            positions.push(position);
+        }
+
+        positions
     }
 
     pub fn get(&mut self, account_id: AccountId, index: u32) -> Option<Position> {
-        let positions_length = self.get_length(account_id);
+        let position = self.positions.get((account_id, index));
+        position
+    }
 
-        if index < positions_length {
-            let positions = self.get_all(account_id);
-            positions.get(index as usize).cloned()
-        } else {
-            None
-        }
+    pub fn get_length(&self, account_id: AccountId) -> u32 {
+        let positions_length = self.positions_length.get(account_id).unwrap_or(0);
+        positions_length
     }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
 
-    fn get_length(&self, account_id: AccountId) -> u32 {
-        let positions = self.get_value(account_id);
-        positions.0
+    #[ink::test]
+    fn test_add() {
+        let positions = &mut Positions::default();
+        let account_id = AccountId::from([0x01; 32]);
+        let position = Position::default();
+
+        positions.add(account_id, position);
+
+        assert_eq!(positions.get(account_id, 0), Some(position));
+        assert_eq!(positions.get(account_id, 1), None);
```

Comment:
Initialization is the first case, and adding the next element is the next case to test. So, add here another `positions.add`
(I know that next tests may have more than one addition, but it is still worth testing it in the tests dedicated to adding)